store/sqlstore/plugin_store.go

```diff
@@ -60,12 +60,7 @@ func (ps SqlPluginStore) SaveOrUpdate(kv *model.PluginKeyValue) (*model.PluginKe
 		} else if rowsAffected == 0 {
 			// No rows were affected by the update, so let's try an insert
 			if err := ps.GetMaster().Insert(kv); err != nil {
-				// If the error is from unique constraints violation, it's the result of a
-				// valid race and we can report success. Otherwise we have a real error and
-				// need to return it
-				if !IsUniqueConstraintError(err, []string{"PRIMARY", "PluginId", "Key", "PKey", "pkey"}) {
-					return nil, model.NewAppError("SqlPluginStore.SaveOrUpdate", "store.sql_plugin_store.save.app_error", nil, err.Error(), http.StatusInternalServerError)
-				}
+				return nil, model.NewAppError("SqlPluginStore.SaveOrUpdate", "store.sql_plugin_store.save.app_error", nil, err.Error(), http.StatusInternalServerError)
```

Comment:
I'm a bit confused about the reasoning given in the commit message... If the update didn't work, there wasn't a key. If we then try the insert and there was a constraint violation, the row must have been inserted after the update. It seems like returning false is the right thing to do. We "did not successfully save or update the value", but it wasn't an internal server error, right?

Reply:
You're right that `http.StatusInternalServerError` isn't really the right error code to use here. Perhaps `http.StatusBadRequest`, fully recognizing that we're abusing the status codes altogether here. Thoughts?

Note too we can't actually return `false` from `SaveOrUpdate` like we can (and still do) in `CompareAndSet` under the same circumstances. So I think an error is the right way to model the failure in this case, vs. swallowing and pretending like it was successful.

Reply:
In that case, can we use `StatusBadRequest` -- that makes more sense to me. :)

Reply:
Fixed. I note that I don't explicitly have coverage for this section of the code. At one point in my refactoring, I did -- launched a bunch of inserters and deleters in threads and raced them till I was happy with the outcome, but ran into deadlock/serialization errors from MySQL, which in turn necessitated retry logic, but tuning the right number of retries got tricky... so I ripped that out in favour of forcing the client to deal with it.

Reply:
True, that _would_ be a pain to test. We tested the workflow contention/retry logic by having an action accidentally hold onto the lock for a very long time... :)